{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iDmXpOtYz5y9"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import random\n",
        "\n",
        "from typing import List, Tuple, Dict\n",
        "\n",
        "filename = \"data/adjusted.csv\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6tMxb1s1z5zA",
        "outputId": "19195801-e97e-4d5a-d744-ed934ce3efa6"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>role</th>\n",
              "      <th>task</th>\n",
              "      <th>general_instructions</th>\n",
              "      <th>user_question</th>\n",
              "      <th>normalized_fitness</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Telomere Biology Specialist</td>\n",
              "      <td>Verify the change in {gene2} when {gene1} is m...</td>\n",
              "      <td>Reveal the molecular link between {gene1} and ...</td>\n",
              "      <td>\"Indicate the causal relationship or function ...</td>\n",
              "      <td>0.214286</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Biomedical Analytics Expert</td>\n",
              "      <td>Probe into the depth of the KEGG Pathway Datab...</td>\n",
              "      <td>Employ the following denointments when express...</td>\n",
              "      <td>Using 'activation', 'inhibition', and 'phospho...</td>\n",
              "      <td>0.357143</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Gene Ontology Specialist</td>\n",
              "      <td>Identify the nature of interaction between {ge...</td>\n",
              "      <td>Capture the connection between {gene1} and {ge...</td>\n",
              "      <td>\"Indicate the causal relationship or function ...</td>\n",
              "      <td>0.214286</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Virologist Specializing in RNA Viruses</td>\n",
              "      <td>Unravel the interplay between {gene1} and {gen...</td>\n",
              "      <td>Use a explicit phrase such as \"activation\", \"i...</td>\n",
              "      <td>Clearly identify the interrelation between {ge...</td>\n",
              "      <td>0.571429</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Nutrigenomics Specialist</td>\n",
              "      <td>Your duty is to find how {gene1} might affect ...</td>\n",
              "      <td>Provide an indisputable resolution using the t...</td>\n",
              "      <td>Based on the KEGG Pathway Database, what is th...</td>\n",
              "      <td>1.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>95</th>\n",
              "      <td>Immunogeneticist involved in ABO genetics</td>\n",
              "      <td>Through the KEGG Pathway Database, determine h...</td>\n",
              "      <td>Monitor, and release data, relating {gene1} an...</td>\n",
              "      <td>Referring to the KEGG Pathway Database, what i...</td>\n",
              "      <td>0.571429</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>96</th>\n",
              "      <td>Genetic Immunologist</td>\n",
              "      <td>Unravel the interplay between {gene1} and {gen...</td>\n",
              "      <td>Use a explicit phrase such as \"activation\", \"i...</td>\n",
              "      <td>\"Please report the interaction between {gene1}...</td>\n",
              "      <td>0.500000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>97</th>\n",
              "      <td>Virologist Specializing in RNA Viruses</td>\n",
              "      <td>Illustrate the mechanism of interaction betwee...</td>\n",
              "      <td>Specify the peculiar bio chemical interaction ...</td>\n",
              "      <td>According to the KEGG Pathway Database analysi...</td>\n",
              "      <td>0.500000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>98</th>\n",
              "      <td>Biomedical Analytics Expert</td>\n",
              "      <td>Develop a comprehensive understanding of how {...</td>\n",
              "      <td>VO favorites enclosed state citations often br...</td>\n",
              "      <td>Clearly identify the interrelation between {ge...</td>\n",
              "      <td>0.285714</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>99</th>\n",
              "      <td>Next Generation Sequencing Analyst</td>\n",
              "      <td>Probe into the depth of the KEGG Pathway Datab...</td>\n",
              "      <td>Employ the following denointments when express...</td>\n",
              "      <td>Using \"activation\", \"inhibition\", \"phosphoryla...</td>\n",
              "      <td>0.357143</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>100 rows Ã— 5 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                         role  \\\n",
              "0                 Telomere Biology Specialist   \n",
              "1                 Biomedical Analytics Expert   \n",
              "2                    Gene Ontology Specialist   \n",
              "3      Virologist Specializing in RNA Viruses   \n",
              "4                    Nutrigenomics Specialist   \n",
              "..                                        ...   \n",
              "95  Immunogeneticist involved in ABO genetics   \n",
              "96                       Genetic Immunologist   \n",
              "97     Virologist Specializing in RNA Viruses   \n",
              "98                Biomedical Analytics Expert   \n",
              "99         Next Generation Sequencing Analyst   \n",
              "\n",
              "                                                 task  \\\n",
              "0   Verify the change in {gene2} when {gene1} is m...   \n",
              "1   Probe into the depth of the KEGG Pathway Datab...   \n",
              "2   Identify the nature of interaction between {ge...   \n",
              "3   Unravel the interplay between {gene1} and {gen...   \n",
              "4   Your duty is to find how {gene1} might affect ...   \n",
              "..                                                ...   \n",
              "95  Through the KEGG Pathway Database, determine h...   \n",
              "96  Unravel the interplay between {gene1} and {gen...   \n",
              "97  Illustrate the mechanism of interaction betwee...   \n",
              "98  Develop a comprehensive understanding of how {...   \n",
              "99  Probe into the depth of the KEGG Pathway Datab...   \n",
              "\n",
              "                                 general_instructions  \\\n",
              "0   Reveal the molecular link between {gene1} and ...   \n",
              "1   Employ the following denointments when express...   \n",
              "2   Capture the connection between {gene1} and {ge...   \n",
              "3   Use a explicit phrase such as \"activation\", \"i...   \n",
              "4   Provide an indisputable resolution using the t...   \n",
              "..                                                ...   \n",
              "95  Monitor, and release data, relating {gene1} an...   \n",
              "96  Use a explicit phrase such as \"activation\", \"i...   \n",
              "97  Specify the peculiar bio chemical interaction ...   \n",
              "98  VO favorites enclosed state citations often br...   \n",
              "99  Employ the following denointments when express...   \n",
              "\n",
              "                                        user_question  normalized_fitness  \n",
              "0   \"Indicate the causal relationship or function ...            0.214286  \n",
              "1   Using 'activation', 'inhibition', and 'phospho...            0.357143  \n",
              "2   \"Indicate the causal relationship or function ...            0.214286  \n",
              "3   Clearly identify the interrelation between {ge...            0.571429  \n",
              "4   Based on the KEGG Pathway Database, what is th...            1.000000  \n",
              "..                                                ...                 ...  \n",
              "95  Referring to the KEGG Pathway Database, what i...            0.571429  \n",
              "96  \"Please report the interaction between {gene1}...            0.500000  \n",
              "97  According to the KEGG Pathway Database analysi...            0.500000  \n",
              "98  Clearly identify the interrelation between {ge...            0.285714  \n",
              "99  Using \"activation\", \"inhibition\", \"phosphoryla...            0.357143  \n",
              "\n",
              "[100 rows x 5 columns]"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "def prepare_candidates_map() -> Dict[str, List[str]]:\n",
        "    \"\"\"\n",
        "    Prepare a map of candidates for different categories.\n",
        "\n",
        "    Returns:\n",
        "        candidates_map (Dict[str, List[str]]): A dictionary containing lists of candidates for different categories.\n",
        "            The categories include 'role', 'task', 'general_instructions', and 'user_question'.\n",
        "    \"\"\"\n",
        "\n",
        "    roles_df = pd.read_excel('roles aims/roles.xlsx')\n",
        "    general_instructions_df = pd.read_excel('roles aims/descriptions.xlsx')\n",
        "    user_questions_df = pd.read_excel('roles aims/question.xlsx')\n",
        "    task_df = pd.read_excel('roles aims/aims.xlsx')\n",
        "\n",
        "    roles_candidates = roles_df[\"roles\"].tolist()\n",
        "    general_instructions_candidates = general_instructions_df[\"general_instructions\"].tolist()\n",
        "    user_questions_candidates = user_questions_df[\"user_questions\"].tolist()\n",
        "    task_candidates = task_df[\"aims\"].tolist()\n",
        "\n",
        "    candidates_map = {\n",
        "        \"role\": roles_candidates,\n",
        "        \"task\": task_candidates,\n",
        "        \"general_instructions\": general_instructions_candidates,\n",
        "        \"user_question\": user_questions_candidates\n",
        "    }\n",
        "    return candidates_map\n",
        "\n",
        "\n",
        "def mutate_population(candidates_map: Dict[str, List[str]],\n",
        "                      population_df: pd.DataFrame,\n",
        "                      mutation_cols: pd.DataFrame):\n",
        "    \"\"\"\n",
        "    Mutates the population dataframe by randomly selecting a mutation column and assigning a random value from the\n",
        "    corresponding candidate values to each row in that column.\n",
        "\n",
        "    Args:\n",
        "        candidates_map (Dict[List[str]]): A dictionary mapping mutation columns to their corresponding candidate values.\n",
        "        population_df (pd.DataFrame): The population dataframe to be mutated.\n",
        "        mutation_cols (pd.DataFrame): The columns in the population dataframe that can be mutated.\n",
        "\n",
        "    Returns:\n",
        "        pd.DataFrame: The mutated population dataframe.\n",
        "    \"\"\"\n",
        "    picked_mutation_col = random.choice(mutation_cols)\n",
        "    candidate_values = candidates_map[picked_mutation_col]\n",
        "    population_df[picked_mutation_col] = [random.choice(candidate_values) for i in range(len(population_df))]\n",
        "    return population_df\n",
        "\n",
        "\n",
        "\n",
        "population_df = pd.read_csv(filename, encoding='latin1')\n",
        "candidates_map = prepare_candidates_map()\n",
        "\n",
        "mutation_cols = [\"role\", \"task\", \"general_instructions\", \"user_question\"]\n",
        "new_population_df = mutate_population(candidates_map, population_df, mutation_cols=mutation_cols)\n",
        "\n",
        "new_population_df"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "venv_pl",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.16"
    },
    "colab": {
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}